src/dag.rs

```diff
@@ -1,11 +1,155 @@
+        /// The document with the mismatched index
+        document: Cid,
+        /// The path up until the mismatched index
+        path: SlashedPath,
+        /// The index in original path
+        index: usize,
+        /// Total number of elements found
+        elements: usize,
+    },
+
+    /// Path attempted to resolve through for example string or integer.
+    #[error("tried to resolve through object that had no links")]
+    NoLinks(Cid, SlashedPath),
+
+    /// Path attempted to resolve through a property, index or link which did not exist.
+    #[error("no link named {:?} under {0}", .1.iter().last().unwrap())]
+    NotFound(Cid, SlashedPath),
+}
+
+#[derive(Debug, Error)]
+pub enum UnexpectedResolved {
+    #[error("path resolved to unexpected type of document: {:?} or {}", .0, .1.cid())]
+    UnexpectedCodec(cid::Codec, ResolvedNode),
+    #[error("path did not resolve to a block on {}", .0.cid())]
+    NonBlock(ResolvedNode),
+}
+
+/// Used internally before translating to ResolveError at the top level by using the IpfsPath.
+#[derive(Debug)]
+enum RawResolveLocalError {
+    Loading(Cid, crate::Error),
+    UnsupportedDocument(Cid, Box<dyn StdError + Send + Sync + 'static>),
+    ListIndexOutOfRange {
+        document: Cid,
+        segment_index: usize,
+        index: usize,
+        elements: usize,
+    },
+    InvalidIndex {
+        document: Cid,
+        segment_index: usize,
+    },
+    NoLinks {
+        document: Cid,
+        segment_index: usize,
+    },
+    NotFound {
+        document: Cid,
+        segment_index: usize,
+    },
+}
+
+impl RawResolveLocalError {
+    /// When resolving through multiple documents the local resolving functions `resolve_local_ipld`
+    /// and `resolve_local_dagpb` return document local indices; need to bump the indices with the
+    /// amount of already matched segments in previous documents for the path.
+    fn add_starting_point_in_path(&mut self, start: usize) {
+        use RawResolveLocalError::*;
+        match self {
+            ListIndexOutOfRange {
+                ref mut segment_index,
+                ..
+            }
+            | InvalidIndex {
+                ref mut segment_index,
+                ..
+            }
+            | NoLinks {
+                ref mut segment_index,
+                ..
+            }
+            | NotFound {
+                ref mut segment_index,
+                ..
+            } => {
+                // NOTE: this is the **index** compared to the number of segments matched **count**
+                // from `resolve_local` Ok return value.
+                *segment_index += start;
+            }
+            _ => {}
+        }
+    }
+
+    /// Use the given [`IpfsPath`] to create the truncated [`SlashedPath`] and convert into
+    /// [`ResolveError`]. The path is truncated so that the last segment is the one which failed to
+    /// match. No reason it could also be just signified with an index.
```

Comment:
Not sure if that's the intention here, but "no reason [why] X" should either be followed by a negative (e.g. "couldn't") or something positive, but _stricter_ than "could", e.g. "must"; also note the reordering of "just"
```suggestion
    /// match. No reason it couldn't also be signified with just an index.
```